lib/spack/spack/fetch_strategy.py

```diff
@@ -108,9 +108,18 @@ def __init__(self, **kwargs):
         self.cache_enabled = not kwargs.pop('no_cache', False)
 
     # Subclasses need to implement these methods
+    def clone(self):
```

Comment:
It needs a new method because the fetcher for git (for example) makes optimizations that developers won’t want (like only doing a shallow clone). For git, the methods really are different. I plan to implement the clone method for other fetchers, and to implement the fetch method in terms of the clone method where feasible (which I think should be most of them besides git).
I do need to add a NotImplementedError in the meantime.

The stage currently hardcodes a lot, including the path. I could potentially reimplement it using a stage, but it would still need to bypass the `fetch` method and change the path from `{name}-{version}-{hash}` to whatever the user path requested (default {name})

Reply:
> the fetcher for git (for example) makes optimizations that developers won’t want (like only doing a shallow clone).

Git versions can specify `get_full_repo` which is checked in the GitFetchStrategy - do you think it would be possible to make use of that?

You could use the stage for fetching/aggregating all required resources, then copy the `source_path` to whatever destination you wanted to (i.e. use the `Stage` as an intermediate location).